ldif.go

```diff
@@ -0,0 +1,256 @@
+package ldap
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	// "os"
+	"strconv"
+)
+
+// A basic LDIF parser. This one does currently just support LDIFs like
+// they're generated by tools like ldapsearch(1) / slapcat(8). Change
+// records are not supported.
+type LDIF struct {
+	RelaxedParser bool // ignore parsing errors in a line
+	Entries       []*Entry
+	curLine       int
+	version       int
+	changeType    string
+}
+
+var CR byte = '\x0D'
+var LF byte = '\x0A'
+var SEP string = string([]byte{CR, LF})
+var Comment byte = '#'
+var SPACE byte = ' '
+var SPACES = string(SPACE)
+
+func (l *LDIF) newError(msg string) error {
+	return errors.New(fmt.Sprintf("error on line %d: %s\n", l.curLine, msg))
+}
+
+// Parses the LDIF, the caller is responsible for closing the io.Reader if that's
+// needed. This Parser may be called several times with different io.Readers to
+// combine the files.
+func (l *LDIF) Parse(r io.Reader) (err error) {
+	if r == nil {
+		return errors.New("No reader present")
+	}
+	reader := bufio.NewReader(r)
+	l.curLine = 0
+	l.changeType = ""
+	l.version = 0
+
+	var lines [][]byte
+	var line, nextLine []byte
+
+	for {
+		l.curLine++
+		nextLine, err = reader.ReadBytes(LF)
+		nextLine = bytes.TrimRight(nextLine, SEP)
+		// fmt.Fprintf(os.Stderr, "NEXT=>%s<\n", nextLine)
+		switch err {
+		case nil, io.EOF:
+			switch len(nextLine) {
+			case 0:
+				if len(line) == 0 && err == io.EOF {
+					return nil
+				}
+				lines = append(lines, line)
+				entry, perr := l.parseEntry(lines)
+				if perr != nil {
+					if !l.RelaxedParser {
+						return l.newError(perr.Error())
+					}
+				}
+				l.Entries = append(l.Entries, entry)
+				line = []byte{}
+				lines = [][]byte{}
+				if err == io.EOF {
+					return nil
+				}
+			default:
+				switch nextLine[0] {
+				case Comment:
+					continue
+				case SPACE:
```

Comment:
I'm not familiar enough with the ldif schema to know... is a single leading space enough to determine this?